src/gafaelfawr/handlers/api.py

```diff
@@ -0,0 +1,241 @@
+) -> APILoginResponse:
+    return APILoginResponse(csrf=context.state.csrf)
+
+
+@router.get(
+    "/token-info",
+    response_model=TokenInfo,
+    response_model_exclude_none=True,
+    responses={404: {"description": "Token not found"}},
+)
+async def get_token_info(
+    auth_data: TokenData = Depends(authenticate),
+    context: RequestContext = Depends(context_dependency),
+) -> TokenInfo:
+    token_service = context.factory.create_token_service()
+    info = token_service.get_info(auth_data.token.key)
+    if not info:
+        msg = "Token found in Redis but not database"
+        context.logger.warning(msg)
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail={"type": "invalid_token", "msg": msg},
+        )
+    else:
+        return info
+
+
+@router.get("/user-info", response_model=TokenUserInfo)
+async def get_user_info(
+    auth_data: TokenData = Depends(authenticate),
+) -> TokenUserInfo:
+    return auth_data
+
+
+@router.get(
+    "/users/{username}/tokens",
+    response_model=List[TokenInfo],
+    response_model_exclude_none=True,
+)
+async def get_tokens(
+    username: str,
+    auth_data: TokenData = Depends(authenticate_session),
+    context: RequestContext = Depends(context_dependency),
+) -> List[TokenInfo]:
+    token_service = context.factory.create_token_service()
+    return token_service.list_tokens(auth_data, username)
+
+
+@router.post(
+    "/users/{username}/tokens",
+    status_code=201,
+    dependencies=[Depends(verify_csrf)],
+)
+async def post_tokens(
+    username: str,
+    token_request: UserTokenRequest,
+    response: Response,
+    auth_data: TokenData = Depends(authenticate_session),
+    context: RequestContext = Depends(context_dependency),
+) -> NewToken:
+    token_service = context.factory.create_token_service()
+    token_params = token_request.dict(exclude_unset=True)
+    if "expires" not in token_params or token_params["expires"] is None:
+        token_params["no_expire"] = True
+    try:
+        token = await token_service.create_user_token(
+            auth_data, username, **token_params
+        )
+    except DuplicateTokenNameError as e:
+        raise HTTPException(
+            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            detail={
+                "loc": ["body", "token_name"],
+                "type": "duplicate_token_name",
+                "msg": str(e),
+            },
+        )
+    token_url = f"/auth/api/v1/users/{username}/tokens/{token.key}"
+    response.headers["Location"] = token_url
+    return NewToken(token=str(token))
+
+
+@router.get(
+    "/users/{username}/tokens/{key}",
+    response_model=TokenInfo,
+    response_model_exclude_none=True,
+)
+async def get_token(
+    username: str,
+    key: str,
+    auth_data: TokenData = Depends(authenticate_session),
+    context: RequestContext = Depends(context_dependency),
+) -> TokenInfo:
+    if username != auth_data.username:
```

Comment:
Yeah, agreed, it's a bit of a layering violation because I was trying too hard for code reuse. I'll add a new service layer method that does the authorization check instead.

Correct, admin impersonation isn't implemented yet. Sorry, I should have noted that explicitly.